File: kh-sim/shared/physics/kh_physics.py
"""
kh_physics.py — Kelvin-Helmholtz instability reference implementation
======================================================================
2D incompressible Navier-Stokes, vorticity-streamfunction formulation.
Doubly periodic domain, pseudo-spectral Poisson solver, RK4 time stepping.

This is the CANONICAL reference. All backend implementations (Rust, Scala,
C++, Fortran, Pascal) must validate their output against this module.

See: shared/physics/KH-PHYSICS.md for full physics documentation.
See: shared/api/openapi.yaml for REST API contract.

Usage:
    python kh_physics.py                        # run reference smoke test
    python kh_physics.py --nx 128 --ny 64 --steps 200 --re 1000
"""

import numpy as np
import json
import time
import argparse
from dataclasses import dataclass, field, asdict
from typing import Optional, List


# ── Data structures (mirrors REST API schema) ──────────────────────────────────

@dataclass
class SimulationRequest:
    grid_nx:               int   = 128
    grid_ny:               int   = 64
    domain_lx:             float = 1.0
    domain_ly:             float = 0.5
    dt:                    float = 0.001
    steps:                 int   = 100
    reynolds_number:       float = 1000.0
    velocity_shear:        float = 1.0
    perturbation_amplitude:float = 0.01
    perturbation_mode:     int   = 2
    initial_omega:         Optional[List[float]] = None  # flattened Nx*Ny


@dataclass
class Diagnostics:
    kinetic_energy:  float = 0.0
    enstrophy:       float = 0.0
    max_vorticity:   float = 0.0
    divergence_rms:  float = 0.0


@dataclass
class SimulationResult:
    backend:          str   = "python-numpy"
    language:         str   = "Python"
    steps_completed:  int   = 0
    t_final:          float = 0.0
    grid_nx:          int   = 0
    grid_ny:          int   = 0
    u_velocity:       List[float] = field(default_factory=list)
    v_velocity:       List[float] = field(default_factory=list)
    vorticity:        List[float] = field(default_factory=list)
    pressure:         List[float] = field(default_factory=list)  # streamfunction ψ
    diagnostics:      Diagnostics = field(default_factory=Diagnostics)
    compute_time_ms:  float = 0.0


# ── Physics kernels ────────────────────────────────────────────────────────────

def _wavenumbers(nx: int, ny: int, dx: float, dy: float):
    """
    Return 2D wavenumber arrays for rfft2 output shape (nx, ny//2+1).
    rfft2 applies full FFT along axis-0 (kx) and real FFT along axis-1 (ky).
    """
    kx = 2.0 * np.pi * np.fft.fftfreq(nx, d=dx)    # full FFT axis-0: shape (nx,)
    ky = 2.0 * np.pi * np.fft.rfftfreq(ny, d=dy)   # real FFT axis-1: shape (ny//2+1,)
    KX, KY = np.meshgrid(kx, ky, indexing='ij')    # shape (nx, ny//2+1) ✓
    return KX, KY


def initial_conditions(nx: int, ny: int, lx: float, ly: float,
                       U0: float, delta: float, amp: float, mode: int
                       ) -> np.ndarray:
    """
    Compute initial vorticity field analytically.

    IC velocity:
        u(x,y) = U0 * tanh((y - Ly/2) / delta)
        v(x,y) = amp * sin(2*pi*mode*x / Lx)

    Vorticity ω = dv/dx - du/dy:
        = amp*(2*pi*mode/Lx)*cos(2*pi*mode*x/Lx)  [from v]
        + U0/delta * sech²((y-Ly/2)/delta)         [from -du/dy, note sign]
    """
    dx = lx / nx
    dy = ly / ny
    x = np.arange(nx) * dx
    y = np.arange(ny) * dy
    X, Y = np.meshgrid(x, y, indexing='ij')

    # Vorticity from shear layer (positive: upper layer faster)
    omega_shear = U0 / delta / np.cosh((Y - ly / 2.0) / delta) ** 2

    # Vorticity perturbation from sinusoidal v-perturbation
    omega_pert = amp * (2.0 * np.pi * mode / lx) * np.cos(2.0 * np.pi * mode * X / lx)

    return omega_shear + omega_pert


def solve_poisson(omega_hat: np.ndarray, KX: np.ndarray, KY: np.ndarray,
                  nx: int, ny: int) -> np.ndarray:
    """
    Spectral Poisson solve: ∇²ψ = -ω  →  ψ̂ = ω̂ / (kx²+ky²)

    Returns psi_hat (rfft2 coefficients).
    Mean mode (kx=ky=0) set to zero: fixes ψ_mean = 0.
    """
    K2 = KX ** 2 + KY ** 2
    K2[0, 0] = 1.0          # avoid divide-by-zero; corrected below
    psi_hat = omega_hat / K2
    psi_hat[0, 0] = 0.0     # zero mean streamfunction
    return psi_hat


def velocity_from_psi_hat(psi_hat: np.ndarray, KX: np.ndarray, KY: np.ndarray,
                           nx: int, ny: int) -> tuple:
    """
    Recover velocity from streamfunction via spectral differentiation.
        u =  dψ/dy  →  û = i·ky·ψ̂
        v = -dψ/dx  →  v̂ = -i·kx·ψ̂
    """
    u = np.fft.irfft2(1j * KY * psi_hat, s=(nx, ny))
    v = np.fft.irfft2(-1j * KX * psi_hat, s=(nx, ny))
    return u, v


def vorticity_rhs(omega: np.ndarray, u: np.ndarray, v: np.ndarray,
                  nu: float, KX: np.ndarray, KY: np.ndarray,
                  nx: int, ny: int) -> np.ndarray:
    """
    RHS of vorticity transport equation (spectral):
        f(ω) = -u·dω/dx - v·dω/dy + ν·∇²ω
    """
    omega_hat = np.fft.rfft2(omega)
    K2 = KX ** 2 + KY ** 2

    domega_dx = np.fft.irfft2(1j * KX * omega_hat, s=(nx, ny))
    domega_dy = np.fft.irfft2(1j * KY * omega_hat, s=(nx, ny))
    laplacian_omega = np.fft.irfft2(-K2 * omega_hat, s=(nx, ny))

    return -u * domega_dx - v * domega_dy + nu * laplacian_omega


def rk4_step(omega: np.ndarray, nu: float,
             KX: np.ndarray, KY: np.ndarray,
             nx: int, ny: int, dt: float) -> np.ndarray:
    """Single RK4 time step for vorticity field."""

    def step_with_omega(w):
        w_hat = np.fft.rfft2(w)
        psi_hat = solve_poisson(w_hat, KX, KY, nx, ny)
        u, v = velocity_from_psi_hat(psi_hat, KX, KY, nx, ny)
        return vorticity_rhs(w, u, v, nu, KX, KY, nx, ny)

    k1 = step_with_omega(omega)
    k2 = step_with_omega(omega + 0.5 * dt * k1)
    k3 = step_with_omega(omega + 0.5 * dt * k2)
    k4 = step_with_omega(omega + dt * k3)

    return omega + (dt / 6.0) * (k1 + 2.0 * k2 + 2.0 * k3 + k4)


# ── Main simulation entry point ────────────────────────────────────────────────

def simulate(req: SimulationRequest) -> SimulationResult:
    """
    Run KH instability simulation. Returns SimulationResult matching REST schema.
    """
    t_start = time.perf_counter()

    nx, ny   = req.grid_nx, req.grid_ny
    lx, ly   = req.domain_lx, req.domain_ly
    dt       = req.dt
    steps    = req.steps
    nu       = req.velocity_shear / req.reynolds_number
    delta    = 0.05 * ly

    dx, dy   = lx / nx, ly / ny
    KX, KY   = _wavenumbers(nx, ny, dx, dy)

    # Initialise vorticity
    if req.initial_omega is not None:
        omega = np.array(req.initial_omega, dtype=float).reshape(nx, ny)
    else:
        omega = initial_conditions(nx, ny, lx, ly,
                                   U0=req.velocity_shear,
                                   delta=delta,
                                   amp=req.perturbation_amplitude,
                                   mode=req.perturbation_mode)

    # Time loop
    t = 0.0
    for _ in range(steps):
        omega = rk4_step(omega, nu, KX, KY, nx, ny, dt)
        t += dt

    # Final field recovery
    omega_hat = np.fft.rfft2(omega)
    psi_hat   = solve_poisson(omega_hat, KX, KY, nx, ny)
    psi       = np.fft.irfft2(psi_hat, s=(nx, ny))
    u, v      = velocity_from_psi_hat(psi_hat, KX, KY, nx, ny)

    # Diagnostics
    ke         = 0.5 * float(np.mean(u ** 2 + v ** 2))
    enstrophy  = 0.5 * float(np.mean(omega ** 2))
    max_vort   = float(np.max(np.abs(omega)))
    # Divergence should be machine-zero for spectral method
    div        = float(np.sqrt(np.mean(
        (np.fft.irfft2(1j * KX * np.fft.rfft2(u), s=(nx, ny)) +
         np.fft.irfft2(1j * KY * np.fft.rfft2(v), s=(nx, ny))) ** 2
    )))

    elapsed_ms = (time.perf_counter() - t_start) * 1000.0

    return SimulationResult(
        backend          = "python-numpy",
        language         = "Python",
        steps_completed  = steps,
        t_final          = round(t, 10),
        grid_nx          = nx,
        grid_ny          = ny,
        u_velocity       = u.flatten().tolist(),
        v_velocity       = v.flatten().tolist(),
        vorticity        = omega.flatten().tolist(),
        pressure         = psi.flatten().tolist(),
        diagnostics      = Diagnostics(
            kinetic_energy = ke,
            enstrophy      = enstrophy,
            max_vorticity  = max_vort,
            divergence_rms = div,
        ),
        compute_time_ms  = round(elapsed_ms, 2),
    )


# ── CLI / smoke test ───────────────────────────────────────────────────────────

def _parse_args():
    p = argparse.ArgumentParser(description="KH instability reference solver")
    p.add_argument("--nx",    type=int,   default=64,    help="Grid points x")
    p.add_argument("--ny",    type=int,   default=32,    help="Grid points y")
    p.add_argument("--lx",    type=float, default=1.0,   help="Domain length x")
    p.add_argument("--ly",    type=float, default=0.5,   help="Domain length y")
    p.add_argument("--dt",    type=float, default=0.001, help="Time step")
    p.add_argument("--steps", type=int,   default=100,   help="Steps to run")
    p.add_argument("--re",    type=float, default=1000.0,help="Reynolds number")
    p.add_argument("--amp",   type=float, default=0.01,  help="Perturbation amplitude")
    p.add_argument("--mode",  type=int,   default=2,     help="Perturbation mode")
    p.add_argument("--out",   type=str,   default="kh_reference_output.json",
                   help="Output JSON path")
    p.add_argument("--full",  action="store_true",
                   help="Include full field arrays in output (large)")
    return p.parse_args()


if __name__ == "__main__":
    args = _parse_args()

    req = SimulationRequest(
        grid_nx=args.nx, grid_ny=args.ny,
        domain_lx=args.lx, domain_ly=args.ly,
        dt=args.dt, steps=args.steps,
        reynolds_number=args.re,
        perturbation_amplitude=args.amp,
        perturbation_mode=args.mode,
    )

    print(f"KH instability reference simulation")
    print(f"  Grid: {req.grid_nx}x{req.grid_ny} | dt={req.dt} | steps={req.steps} | Re={req.reynolds_number}")
    print(f"  Domain: [{req.domain_lx} x {req.domain_ly}] | amp={req.perturbation_amplitude} | mode={req.perturbation_mode}")
    print()

    result = simulate(req)

    print(f"  t_final          = {result.t_final:.4f}")
    print(f"  kinetic_energy   = {result.diagnostics.kinetic_energy:.6f}")
    print(f"  enstrophy        = {result.diagnostics.enstrophy:.6f}")
    print(f"  max_vorticity    = {result.diagnostics.max_vorticity:.6f}")
    print(f"  divergence_rms   = {result.diagnostics.divergence_rms:.2e}  (should be < 1e-10)")
    print(f"  compute_time     = {result.compute_time_ms:.1f} ms")

    # Save reference output (strip large arrays unless --full)
    out_dict = asdict(result)
    if not args.full:
        for fld in ("u_velocity", "v_velocity", "vorticity", "pressure"):
            vals = out_dict[fld]
            out_dict[fld] = {
                "length":   len(vals),
                "first_10": vals[:10],
                "last_10":  vals[-10:],
                "min":      float(min(vals)),
                "max":      float(max(vals)),
                "mean":     float(sum(vals) / len(vals)),
            }

    with open(args.out, "w") as f:
        json.dump(out_dict, f, indent=2)

    print(f"\n  Reference output -> {args.out}")
    print("\n  [PASS] Reference simulation complete.")

File: kh-sim/backends/pascal/src/kh_physics.pas
unit kh_physics;
{ KH instability physics kernel — Pascal (Free Pascal 3.2+)
  Physics : 2D incompressible Navier-Stokes, vorticity-streamfunction form.
  Numerics : pseudo-spectral (Cooley-Tukey radix-2), RK4.
  Reference: kh-sim/shared/physics/KH-PHYSICS.md
  Canonical: kh-sim/shared/physics/kh_physics.py
}
{$mode objfpc}{$H+}

interface

uses Math, SysUtils, DateUtils;

{ ── Complex arithmetic ───────────────────────────────────────────────────── }
type
  TComplex    = record re, im: Double; end;
  TComplexArr = array of TComplex;
  TRealArr    = array of Double;

operator + (const a, b: TComplex): TComplex; inline;
operator - (const a, b: TComplex): TComplex; inline;
operator * (const a, b: TComplex): TComplex; inline;
operator * (const a: Double; const b: TComplex): TComplex; inline;
operator * (const a: TComplex; const b: Double): TComplex; inline;
operator / (const a: TComplex; const b: Double): TComplex; inline;

function CmplxOf(re, im: Double): TComplex; inline;
function CmplxFromReal(const r: TRealArr; n: Integer): TComplexArr;
function RealPart(const c: TComplexArr; n: Integer): TRealArr;

{ ── Simulation API ───────────────────────────────────────────────────────── }
type
  TSimRequest = record
    nx, ny, steps, mode: Integer;
    lx, ly, dt, re, u0, amp: Double;
  end;

  TSimResult = record
    u, v, omega, psi: TRealArr;
    ke, enstrophy, max_vort, div_rms, compute_ms: Double;
  end;

function Simulate(const req: TSimRequest): TSimResult;

implementation

{ ── Operator implementations ─────────────────────────────────────────────── }
operator + (const a, b: TComplex): TComplex;
begin Result.re := a.re + b.re; Result.im := a.im + b.im; end;

operator - (const a, b: TComplex): TComplex;
begin Result.re := a.re - b.re; Result.im := a.im - b.im; end;

operator * (const a, b: TComplex): TComplex;
begin
  Result.re := a.re*b.re - a.im*b.im;
  Result.im := a.re*b.im + a.im*b.re;
end;

operator * (const a: Double; const b: TComplex): TComplex;
begin Result.re := a*b.re; Result.im := a*b.im; end;

operator * (const a: TComplex; const b: Double): TComplex;
begin Result.re := a.re*b; Result.im := a.im*b; end;

operator / (const a: TComplex; const b: Double): TComplex;
begin Result.re := a.re/b; Result.im := a.im/b; end;

function CmplxOf(re, im: Double): TComplex;
begin Result.re := re; Result.im := im; end;

function CmplxFromReal(const r: TRealArr; n: Integer): TComplexArr;
var i: Integer;
begin
  SetLength(Result, n);
  for i := 0 to n-1 do begin Result[i].re := r[i]; Result[i].im := 0.0; end;
end;

function RealPart(const c: TComplexArr; n: Integer): TRealArr;
var i: Integer;
begin
  SetLength(Result, n);
  for i := 0 to n-1 do Result[i] := c[i].re;
end;

{ ── 1D in-place Cooley-Tukey FFT (radix-2, DIT) ─────────────────────────── }
{ n must be a power of 2. inverse=True applies 1/n normalisation.             }
procedure FFT1D(var a: TComplexArr; n: Integer; inverse: Boolean);
var
  i, j, bit, len, k: Integer;
  ang: Double;
  wlen, w, u, v: TComplex;
begin
  { bit-reversal permutation }
  j := 0;
  for i := 1 to n-1 do begin
    bit := n shr 1;
    while (j and bit) <> 0 do begin
      j := j xor bit;
      bit := bit shr 1;
    end;
    j := j xor bit;
    if i < j then begin u := a[i]; a[i] := a[j]; a[j] := u; end;
  end;
  { butterfly stages }
  len := 2;
  while len <= n do begin
    ang := 2.0*Pi / len;
    if inverse then ang := -ang;
    wlen := CmplxOf(Cos(ang), Sin(ang));
    i := 0;
    while i < n do begin
      w := CmplxOf(1.0, 0.0);
      for k := 0 to (len shr 1)-1 do begin
        u := a[i+k];
        v := a[i+k+(len shr 1)] * w;
        a[i+k]              := u + v;
        a[i+k+(len shr 1)] := u - v;
        w := w * wlen;
      end;
      i := i + len;
    end;
    len := len shl 1;
  end;
  { normalise inverse }
  if inverse then
    for i := 0 to n-1 do a[i] := a[i] / n;
end;

{ ── 2D FFT: row-then-column (row-major a[i*ny+j]) ───────────────────────── }
procedure FFT2D(var a: TComplexArr; nx, ny: Integer; inverse: Boolean);
var
  i, j: Integer;
  row, col: TComplexArr;
begin
  SetLength(row, ny);
  SetLength(col, nx);
  { FFT along y-axis (rows of length ny) }
  for i := 0 to nx-1 do begin
    for j := 0 to ny-1 do row[j] := a[i*ny+j];
    FFT1D(row, ny, inverse);
    for j := 0 to ny-1 do a[i*ny+j] := row[j];
  end;
  { FFT along x-axis (columns of length nx) }
  for j := 0 to ny-1 do begin
    for i := 0 to nx-1 do col[i] := a[i*ny+j];
    FFT1D(col, nx, inverse);
    for i := 0 to nx-1 do a[i*ny+j] := col[i];
  end;
end;

{ ── Angular wavenumber meshgrid, row-major ───────────────────────────────── }
{ kx[i*ny+j] = 2π*fftfreq(i,nx)/dx,  ky[i*ny+j] = 2π*fftfreq(j,ny)/dy     }
procedure MakeWavenumbers(nx, ny: Integer; dx, dy: Double;
  out kx, ky: TRealArr);
var i, j, fi, fj: Integer;
begin
  SetLength(kx, nx*ny);
  SetLength(ky, nx*ny);
  for i := 0 to nx-1 do begin
    if i <= nx div 2 then fi := i else fi := i - nx;
    for j := 0 to ny-1 do kx[i*ny+j] := 2.0*Pi*fi / (nx*dx);
  end;
  for j := 0 to ny-1 do begin
    if j <= ny div 2 then fj := j else fj := j - ny;
    for i := 0 to nx-1 do ky[i*ny+j] := 2.0*Pi*fj / (ny*dy);
  end;
end;

{ ── Initial conditions: tanh shear + sinusoidal perturbation ─────────────── }
procedure InitConds(nx, ny: Integer; lx, ly, u0, delta, amp: Double;
  mode: Integer; out omega: TRealArr);
var i, j: Integer;
    dx, dy, x, y, z, pert: Double;
begin
  SetLength(omega, nx*ny);
  dx := lx / nx;  dy := ly / ny;
  for i := 0 to nx-1 do begin
    x    := i * dx;
    pert := amp * (2.0*Pi*mode/lx) * Cos(2.0*Pi*mode*x/lx);
    for j := 0 to ny-1 do begin
      y := j * dy;
      z := (y - ly*0.5) / delta;
      omega[i*ny+j] := (u0/delta) / (Cosh(z)*Cosh(z)) + pert;
    end;
  end;
end;

{ ── Spectral Poisson: psi_hat = omega_hat / k², zero mode = 0 ───────────── }
procedure SolvePoisson(const oh: TComplexArr; const kx, ky: TRealArr;
  nx, ny: Integer; out ph: TComplexArr);
var i, j, idx: Integer; k2: Double;
begin
  SetLength(ph, nx*ny);
  for i := 0 to nx-1 do
    for j := 0 to ny-1 do begin
      idx := i*ny+j;
      if (i=0) and (j=0) then
        ph[idx] := CmplxOf(0.0, 0.0)
      else begin
        k2 := kx[idx]*kx[idx] + ky[idx]*ky[idx];
        ph[idx] := oh[idx] / k2;
      end;
    end;
end;

{ ── Velocity from streamfunction ─────────────────────────────────────────── }
{ u_hat = i·ky·ψ̂  →  re=-ky·ψ̂.im, im=+ky·ψ̂.re                           }
{ v_hat = -i·kx·ψ̂ →  re=+kx·ψ̂.im, im=-kx·ψ̂.re                           }
procedure VelFromPsi(const ph: TComplexArr; const kx, ky: TRealArr;
  nx, ny: Integer; out u, v: TRealArr);
var i, j, idx, n: Integer;
    uh, vh: TComplexArr;
begin
  n := nx*ny;
  SetLength(uh, n); SetLength(vh, n);
  for i := 0 to nx-1 do
    for j := 0 to ny-1 do begin
      idx := i*ny+j;
      uh[idx] := CmplxOf(-ky[idx]*ph[idx].im,  ky[idx]*ph[idx].re);
      vh[idx] := CmplxOf( kx[idx]*ph[idx].im, -kx[idx]*ph[idx].re);
    end;
  FFT2D(uh, nx, ny, True);
  FFT2D(vh, nx, ny, True);
  u := RealPart(uh, n);
  v := RealPart(vh, n);
end;

{ ── Vorticity RHS: -u·∂ω/∂x - v·∂ω/∂y + ν·∇²ω ────────────────────────── }
procedure VortRHS(const omega, uin, vin, kx, ky: TRealArr;
  nx, ny: Integer; nu: Double; out rhs: TRealArr);
var
  i, j, idx, n: Integer;
  k2: Double;
  oh, dxs, dys, laps: TComplexArr;
  dox, doy, lap: TRealArr;
begin
  n := nx*ny;
  oh := CmplxFromReal(omega, n);
  FFT2D(oh, nx, ny, False);
  SetLength(dxs, n); SetLength(dys, n); SetLength(laps, n);
  for i := 0 to nx-1 do
    for j := 0 to ny-1 do begin
      idx := i*ny+j;
      k2  := kx[idx]*kx[idx] + ky[idx]*ky[idx];
      { i·kx·ω̂ }  dxs[idx]  := CmplxOf(-kx[idx]*oh[idx].im,  kx[idx]*oh[idx].re);
      { i·ky·ω̂ }  dys[idx]  := CmplxOf(-ky[idx]*oh[idx].im,  ky[idx]*oh[idx].re);
      { -k²·ω̂  }  laps[idx] := oh[idx] * (-k2);
    end;
  FFT2D(dxs,  nx, ny, True);  dox := RealPart(dxs,  n);
  FFT2D(dys,  nx, ny, True);  doy := RealPart(dys,  n);
  FFT2D(laps, nx, ny, True);  lap := RealPart(laps, n);
  SetLength(rhs, n);
  for idx := 0 to n-1 do
    rhs[idx] := -uin[idx]*dox[idx] - vin[idx]*doy[idx] + nu*lap[idx];
end;

{ ── RK4 time step ────────────────────────────────────────────────────────── }
procedure RK4Step(const omega, kx, ky: TRealArr;
  nx, ny: Integer; nu, dt: Double; out omega_out: TRealArr);
var
  n, idx: Integer;
  k1, k2, k3, k4, tmp, u, v: TRealArr;
  oh, ph: TComplexArr;
begin
  n := nx*ny;
  SetLength(tmp, n);

  oh := CmplxFromReal(omega, n); FFT2D(oh, nx, ny, False);
  SolvePoisson(oh, kx, ky, nx, ny, ph);
  VelFromPsi(ph, kx, ky, nx, ny, u, v);
  VortRHS(omega, u, v, kx, ky, nx, ny, nu, k1);

  for idx := 0 to n-1 do tmp[idx] := omega[idx] + 0.5*dt*k1[idx];
  oh := CmplxFromReal(tmp, n); FFT2D(oh, nx, ny, False);
  SolvePoisson(oh, kx, ky, nx, ny, ph);
  VelFromPsi(ph, kx, ky, nx, ny, u, v);
  VortRHS(tmp, u, v, kx, ky, nx, ny, nu, k2);

  for idx := 0 to n-1 do tmp[idx] := omega[idx] + 0.5*dt*k2[idx];
  oh := CmplxFromReal(tmp, n); FFT2D(oh, nx, ny, False);
  SolvePoisson(oh, kx, ky, nx, ny, ph);
  VelFromPsi(ph, kx, ky, nx, ny, u, v);
  VortRHS(tmp, u, v, kx, ky, nx, ny, nu, k3);

  for idx := 0 to n-1 do tmp[idx] := omega[idx] + dt*k3[idx];
  oh := CmplxFromReal(tmp, n); FFT2D(oh, nx, ny, False);
  SolvePoisson(oh, kx, ky, nx, ny, ph);
  VelFromPsi(ph, kx, ky, nx, ny, u, v);
  VortRHS(tmp, u, v, kx, ky, nx, ny, nu, k4);

  SetLength(omega_out, n);
  for idx := 0 to n-1 do
    omega_out[idx] := omega[idx] + (dt/6.0)*(k1[idx] + 2.0*k2[idx] + 2.0*k3[idx] + k4[idx]);
end;

{ ── Main simulate entry point ────────────────────────────────────────────── }
function Simulate(const req: TSimRequest): TSimResult;
var
  nx, ny, n, s, idx: Integer;
  lx, ly, dt, nu, delta, dx, dy: Double;
  omega, omega_next, kx, ky, u, v, psi_r: TRealArr;
  oh, ph, ph2, uh, vh, div_s: TComplexArr;
  ke_sum, ens_sum, max_w, div_sum, n_inv: Double;
  t_start: TDateTime;
begin
  nx := req.nx;  ny := req.ny;  n := nx*ny;
  lx := req.lx;  ly := req.ly;  dt := req.dt;
  nu    := req.u0 / req.re;
  delta := 0.05 * ly;
  dx    := lx / nx;
  dy    := ly / ny;

  MakeWavenumbers(nx, ny, dx, dy, kx, ky);
  InitConds(nx, ny, lx, ly, req.u0, delta, req.amp, req.mode, omega);

  t_start := Now;
  for s := 1 to req.steps do begin
    RK4Step(omega, kx, ky, nx, ny, nu, dt, omega_next);
    omega := omega_next;
  end;
  Result.compute_ms := MilliSecondsBetween(Now, t_start);

  { final psi via ifft(solve_poisson(fft(omega))) }
  oh := CmplxFromReal(omega, n);  FFT2D(oh, nx, ny, False);
  SolvePoisson(oh, kx, ky, nx, ny, ph);
  SetLength(ph2, n);
  for idx := 0 to n-1 do ph2[idx] := ph[idx];
  FFT2D(ph2, nx, ny, True);
  psi_r := RealPart(ph2, n);

  { re-solve ph for velocity (ph was consumed by ifft above) }
  oh := CmplxFromReal(omega, n);  FFT2D(oh, nx, ny, False);
  SolvePoisson(oh, kx, ky, nx, ny, ph);
  VelFromPsi(ph, kx, ky, nx, ny, u, v);

  { diagnostics }
  n_inv := 1.0 / n;
  ke_sum := 0.0;  ens_sum := 0.0;  max_w := 0.0;
  for idx := 0 to n-1 do begin
    ke_sum  := ke_sum  + 0.5*(u[idx]*u[idx] + v[idx]*v[idx]);
    ens_sum := ens_sum + 0.5*omega[idx]*omega[idx];
    if Abs(omega[idx]) > max_w then max_w := Abs(omega[idx]);
  end;

  { divergence rms: div = i·kx·û + i·ky·v̂ }
  uh := CmplxFromReal(u, n);  FFT2D(uh, nx, ny, False);
  vh := CmplxFromReal(v, n);  FFT2D(vh, nx, ny, False);
  SetLength(div_s, n);
  for idx := 0 to n-1 do
    div_s[idx] := CmplxOf(
      -kx[idx]*uh[idx].im - ky[idx]*vh[idx].im,
       kx[idx]*uh[idx].re + ky[idx]*vh[idx].re);
  FFT2D(div_s, nx, ny, True);
  div_sum := 0.0;
  for idx := 0 to n-1 do div_sum := div_sum + div_s[idx].re * div_s[idx].re;

  Result.u          := u;
  Result.v          := v;
  Result.omega      := omega;
  Result.psi        := psi_r;
  Result.ke         := ke_sum  * n_inv;
  Result.enstrophy  := ens_sum * n_inv;
  Result.max_vort   := max_w;
  Result.div_rms    := Sqrt(div_sum * n_inv);
end;

end.
